netbox/dcim/models/device_components.py

```diff
@@ -373,37 +373,67 @@ def connected_endpoint(self, value):
                 "Connected endpoint must be a PowerOutlet or PowerFeed, not {}.".format(type(value))
             )
 
-    def get_power_draw(self):
+    def get_power_draw(self, leg_stats=True):
         """
         Return the allocated and maximum power draw (in VA) and child PowerOutlet count for this PowerPort.
+        If `leg_stats` is True, the `legs` key in the returned dict is populated with a list of per-leg statistics.
         """
+        def get_power_stats(leg=None):
+            """
+            Return tuple of (outlet_count, allocated_draw_total, maximum_draw_total).
+            """
+            # Keep track of all the power ports that have already been processed
+            visited_power_ports = PowerPort.objects.none()
+
+            # Power outlets assigned to the current power port
+            power_outlets = PowerOutlet.objects.filter(power_port=self)
+            if leg is not None:
+                power_outlets = power_outlets.filter(feed_leg=leg)
+
+            # Cannot be cached as it will otherwise not update the per-leg stats when an outlet's leg changes.
+            connected_power_ports = PowerPort.objects.exclude(pk__in=visited_power_ports).filter(
+                _connected_poweroutlet__in=power_outlets).nocache()
```

Comment:
I'd preferably like to avoid `nocache` and simply invalidate this query when the leg is changed on one of the `power_outlets`. Not sure if this is possible in a clean way, though I'm doing a bit of research on this.